clients/drcachesim/tests/trace_interval_analysis_unit_tests.cpp

```diff
@@ -149,63 +340,193 @@ class test_analysis_tool_t : public analysis_tool_t {
+        recorded_snapshot_t *snapshot_two = dynamic_cast<recorded_snapshot_t *>(two);
+        result->component_intervals_ = std::move(snapshot_one->component_intervals_);
+        result->component_intervals_.insert(result->component_intervals_.end(),
+                                            snapshot_two->component_intervals_.begin(),
+                                            snapshot_two->component_intervals_.end());
+        return result;
+    }
+    bool
+    compare_results(std::vector<recorded_snapshot_t *> &one,
+                    std::vector<recorded_snapshot_t> &two)
+    {
+        if (one.size() != two.size())
+            return false;
+        for (int i = 0; i < one.size(); ++i) {
+            std::sort(one[i]->component_intervals_.begin(),
+                      one[i]->component_intervals_.end());
+            std::sort(two[i].component_intervals_.begin(),
+                      two[i].component_intervals_.end());
+            if (!(*one[i] == two[i]))
+                return false;
+        }
         return true;
     }
     bool
-    generate_shard_interval_result(void *shard_data, uint64_t interval_id) override
+    print_interval_results(
+        const std::vector<interval_state_snapshot_t *> &snapshots) override
     {
-        if (shard_data != reinterpret_cast<void *>(kShardData)) {
-            fprintf(stderr, "Invalid shard_data\n");
+        std::vector<recorded_snapshot_t *> recorded_state_snapshots;
+        for (const auto &p : snapshots) {
+            recorded_state_snapshots.push_back(
+                reinterpret_cast<recorded_snapshot_t *>(p));
+        }
+        if (!compare_results(recorded_state_snapshots, expected_state_snapshots_)) {
+            error_string_ = "Unexpected state snapshots";
+            std::cerr << "Expected:\n";
+            for (const auto &snapshot : expected_state_snapshots_)
+                snapshot.print();
+            std::cerr << "Found:\n";
+            for (const auto &snapshot : recorded_state_snapshots)
+                snapshot->print();
             return false;
         }
-        parallel_interval_ends_.push_back(
-            std::make_pair(interval_id, seen_parallel_memrefs_));
         return true;
     }
-
-    std::vector<std::pair<uint64_t, int>> serial_interval_ends_;
-    std::vector<std::pair<uint64_t, int>> parallel_interval_ends_;
+    bool
+    release_interval_snapshot(
+        analysis_tool_t::interval_state_snapshot_t *snapshot) override
+    {
+        delete snapshot;
+        --outstanding_snapshots_;
+        return true;
+    }
+    int
+    get_outstanding_snapshot_count()
+    {
+        return outstanding_snapshots_;
+    }
 
 private:
     int seen_memrefs_;
-    int seen_parallel_memrefs_;
+    std::vector<recorded_snapshot_t> expected_state_snapshots_;
+    int outstanding_snapshots_;
+
+    // Data tracked per shard.
+    struct per_shard_t {
+        memref_tid_t shard_id_;
```

Comment:
style: We use trailing underscores only for class fields

Reply:
Done